File: Asignaturas/Tercer Año/PDOO/Practica/Proyecto_Irrgarten/Proyecto_Irrgarten/SUBIR_A_PRADO_2_PARCIAL_PDOO/VARIOS_DISEÑOS/JAVA_V11-PROYECTO_BASE/P5-java/src/irrgarten/Dice.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

import java.util.Random;
import java.util.ArrayList;

/**
 *
 * @author ismael
 */

/**
 * La clase {@code Dice} proporciona varios métodos estáticos para simular
 * valores aleatorios relacionados con un juego de combate, incluyendo atributos
 * como fuerza, inteligencia, recompensas, probabilidad de resurrección, entre otros.
 * Los valores generados están limitados por constantes establecidas en la clase.
 * Utiliza un generador de números aleatorios para simular estos valores.
 */
public class Dice {
    
    /**
     * Número máximo de usos para armas y escudos.
     */
    private static final int MAX_USES = 5;
    
    /**
     * Valor máximo de inteligencia para jugadores y monstruos.
     */
    private static final float MAX_INTELLIGENCE = 10.0f;
    
    /**
     * Valor máximo de fuerza para jugadores y monstruos.
     */
    private static final float MAX_STRENGTH = 10.0f;
    
    /**
     * Probabilidad de que un jugador sea resucitado en cada turno.
     */
    private static final float RESURRECT_PROB = 0.3f;
    
    /**
     * Máximo número de armas que un jugador puede recibir como recompensa por ganar un combate.
     */
    private static final int WEAPONS_REWARD = 2;
    
    /**
     * Máximo número de escudos que un jugador puede recibir como recompensa por ganar un combate.
     */
    private static final int SHIELDS_REWARD = 3;
    
    /**
     * Máximo número de unidades de salud que un jugador puede recibir como recompensa por ganar un combate.
     */
    private static final int HEALTH_REWARD = 5;
    
    /**
     * Potencia máxima que puede tener un arma.
     */
    private static final int MAX_ATTACK = 3;
    
    /**
     * Potencia máxima que puede tener un escudo.
     */
    private static final int MAX_SHIELD = 2;
    
    /**
     * Generador de números aleatorios utilizado por la clase.
     */
    private static Random generator = new Random();
    
    /**
     * Genera una posición aleatoria dentro del límite máximo dado.
     *
     * @param max el valor máximo (exclusivo) para la posición aleatoria.
     * @return un valor entero aleatorio entre 0 (inclusive) y max (exclusivo).
     */
    static public int randomPos(int max) {
        return generator.nextInt(max);
    }
    
    /**
     * Determina aleatoriamente qué jugador comienza en base al número de jugadores.
     *
     * @param nplayers el número total de jugadores.
     * @return un valor entero entre 0 y nplayers - 1 indicando quién comienza.
     */
    static public int whoStarts(int nplayers) {
        return generator.nextInt(nplayers);
    }
    
    /**
     * Genera un valor aleatorio para la inteligencia, limitado por el máximo establecido.
     *
     * @return un valor flotante aleatorio entre 0.0 y MAX_INTELLIGENCE.
     */
    public static float randomIntelligence(){
        return generator.nextFloat()*MAX_INTELLIGENCE;
    }

    /**
     * Genera un valor aleatorio para la fuerza, limitado por el máximo establecido.
     *
     * @return un valor flotante aleatorio entre 0.0 y MAX_STRENGTH.
     */
    public static float randomStrength(){
        return generator.nextFloat()*MAX_STRENGTH;
    }
    
    /**
     * Determina si un jugador resucitará en un turno, en base a la probabilidad de resurrección.
     *
     * @return {@code true} si el jugador resucita; {@code false} de lo contrario.
     */
    static boolean resurrectPlayer() {
        return generator.nextFloat() <= RESURRECT_PROB;
    }
    
    /**
     * Genera una cantidad aleatoria de armas recibidas como recompensa.
     *
     * @return un valor entero entre 0 y WEAPONS_REWARD .
     */
    static public int weaponsReward(){
        return generator.nextInt(WEAPONS_REWARD+1);
    }
    
    /**
     * Genera una cantidad aleatoria de escudos recibidos como recompensa.
     *
     * @return un valor entero entre 0 y SHIELDS_REWARD.
     */
    static public int shieldsReward(){
        return generator.nextInt(SHIELDS_REWARD+1); //Intervalo cerrado, debemos de incluir el ultimo
   }
    
    /**
     * Genera una cantidad aleatoria de salud recibida como recompensa.
     *
     * @return un valor entero entre 0 y HEALTH_REWARD .
     */
    static public int healthReward(){
        return generator.nextInt(HEALTH_REWARD+1);
    }
    
    /**
     * Genera un valor aleatorio de potencia para un arma.
     *
     * @return un valor flotante entre 0 y MAX_ATTACK .
     */
    static public float weaponPower(){
        return generator.nextFloat()*MAX_ATTACK;
    }
   
    /**
     * Genera un valor aleatorio de potencia para un escudo.
     *
     * @return un valor flotante entre 0 y MAX_SHIELD .
     */
    static public float shieldPower(){
        return generator.nextFloat()*MAX_SHIELD;
    }
    
    /**
     * Genera un número aleatorio de usos restantes para armas o escudos.
     *
     * @return un valor entero entre 0 y MAX_USES .
     */
    static public int usesLeft(){
        return generator.nextInt(MAX_USES+1);
    }
    
    /**
     * Genera una intensidad aleatoria basada en la competencia dada.
     *
     * @param competence el nivel máximo de competencia.
     * @return un valor flotante entre 0.0 y competence.
     */
    static public float intensity(float competence){
        return generator.nextFloat()*competence;
    }
    
    /**
     * Determina si un elemento debe ser descartado en función de sus usos restantes.
     *
     * @param usesLeft el número de usos restantes del elemento.
     * @return {@code true} si el elemento debe ser descartado; {@code false} de lo contrario.
     */
    static public boolean discardElement(int usesLeft) {
        float p = (float)usesLeft / MAX_USES;
        
        return generator.nextFloat() >= p;
    }
    /**
     * 
     */
    /**
     * Determines the next step based on a preferred direction, a list of valid moves, and an intelligence factor.
     *
     * @param preference The preferred direction to move in.
     * @param validMoves A list of valid directions that can be taken.
     * @param intelligence A float value representing the intelligence factor, which influences the likelihood of choosing the preferred direction.
     * @return The chosen direction for the next step, which may be the preferred direction or a random valid move based on the intelligence factor.
     */
    static public Directions nextStep(Directions preference, ArrayList<Directions> validMoves, float intelligence) {
        Directions out;
        if (generator.nextFloat() < (intelligence / MAX_INTELLIGENCE)) {
        out = preference;
        } else {
            out =  validMoves.get(generator.nextInt(validMoves.size()));
        }
         return out;
         
         //Otra implementacion:
         /*Directions toReturn=preference;
        
        if(Dice.randomIntelligence()>intelligence){
            int indice=generator.nextInt(validMoves.size());
            toReturn=validMoves.get(indice);          
        }
        
        return toReturn;*/
    }
}
    

File: Asignaturas/Tercer Año/PDOO/Practica/Proyecto_Irrgarten/Proyecto_Irrgarten/SUBIR_A_PRADO_2_PARCIAL_PDOO/VARIOS_DISEÑOS/PROYECTO_MOVIMIENTOS_PREDEFINIDOS/P5-Ruby/lib/irrgarten/dice.rb
#encoding:utf-8
module Irrgarten
    # Clase que representa un dado para generar valores aleatorios en el juego.
    # Contiene métodos para calcular atributos y valores relacionados con jugadores, monstruos, armas, y escudos.
    class Dice
        @@MAX_USES = 5          # Número máximo de usos de armas y escudos
        @@MAX_INTELLIGENCE = 10.0 # Valor máximo para la inteligencia de jugadores y monstruos
        @@MAX_STRENGTH = 10.0    # Valor máximo para la fuerza de jugadores y monstruos
        @@RESURRECT_PROB = 0.3   # Probabilidad de que un jugador sea resucitado en cada turno
        @@WEAPONS_REWARD = 2     # Número máximo de armas recibidas al ganar un combate
        @@SHIELDS_REWARD = 3     # Número máximo de escudos recibidos al ganar un combate
        @@HEALTH_REWARD = 5      # Número máximo de unidades de salud recibidas al ganar un combate
        @@MAX_ATTACK = 3         # Potencia máxima de las armas
        @@MAX_SHIELD = 2         # Potencia máxima de los escudos

        @@generator = Random.new # Generador de números aleatorios

        # Genera una posición aleatoria hasta un valor máximo.
        #
        # @param max [Integer] Valor máximo para la posición.
        # @return [Integer] Posición aleatoria entre 0 y max - 1.
        def self.randomPos(max)
            @@generator.rand(max.to_i)
        end

        # Determina qué jugador comienza el juego.
        #
        # @param nplayers [Integer] Número de jugadores.
        # @return [Integer] Índice del jugador que empieza, entre 0 y nplayers - 1.
        def self.whoStarts(nplayers)
            @@generator.rand(nplayers.to_i)
        end
        
        # Genera un valor aleatorio para la inteligencia de un jugador o monstruo.
        #
        # @return [Float] Valor de inteligencia entre 0 y MAX_INTELLIGENCE.
        def self.randomIntelligence
            @@generator.rand(@@MAX_INTELLIGENCE.to_f)
        end

        # Genera un valor aleatorio para la fuerza de un jugador o monstruo.
        #
        # @return [Float] Valor de fuerza entre 0 y MAX_STRENGTH.
        def self.randomStrength
            @@generator.rand(@@MAX_STRENGTH.to_f)
        end

        # Determina si un jugador es resucitado según la probabilidad definida.
        #
        # @return [Boolean] `true` si el jugador es resucitado, `false` en caso contrario.
        def self.resurrectPlayer
            @@generator.rand <= @@RESURRECT_PROB
        end

        # Genera un número aleatorio de armas como recompensa tras ganar un combate.
        #
        # @return [Integer] Cantidad de armas recibidas, entre 1 y WEAPONS_REWARD.
        def self.weaponReward
            @@generator.rand(@@WEAPONS_REWARD.to_i+1)
        end  

        # Genera un número aleatorio de escudos como recompensa tras ganar un combate.
        #
        # @return [Integer] Cantidad de escudos recibidos, entre 1 y SHIELDS_REWARD.
        def self.shieldReward
            @@generator.rand(@@SHIELDS_REWARD.to_i+1)
        end

        # Genera una cantidad aleatoria de salud como recompensa tras ganar un combate.
        #
        # @return [Integer] Cantidad de salud recibida, entre 1 y HEALTH_REWARD.
        def self.healthReward
            @@generator.rand(@@HEALTH_REWARD.to_i+1)
        end

        # Genera la potencia de un arma aleatoria.
        #
        # @return [Integer] Potencia del arma, entre 0 y MAX_ATTACK.
        def self.weaponPower
            @@generator.rand(@@MAX_ATTACK.to_f)
        end

        # Genera la potencia de un escudo aleatorio.
        #
        # @return [Integer] Potencia del escudo, entre 0 y MAX_SHIELD.
        def self.shieldPower
            @@generator.rand(@@MAX_SHIELD.to_f)
        end

        # Genera un número aleatorio de usos restantes para un arma o escudo.
        #
        # @return [Integer] Número de usos, entre 1 y MAX_USES.
        def self.usesLeft
            @@generator.rand(@@MAX_USES.to_i+1)
        end 

        # Calcula la intensidad de un ataque o defensa según una competencia.
        #
        # @param competence [Float] Competencia del jugador o monstruo.
        # @return [Float] Intensidad aleatoria entre 0 y el valor de competencia.
        def self.intensity(competence)
            @@generator.rand(competence.to_f)
        end

        # Genera un valor booleano que indica si un elemento debe ser descartado
        #
        # @param usesLeft [Integer] El número de usos restantes del elemento
        # @return [Boolean] `true` si el elemento debe ser descartado, `false` en caso contrario
        def self.discardElement(usesLeft)
            probabilidad=usesLeft.to_f/@@MAX_USES
            return (@@generator.rand >= probabilidad)
        end

        #Otra forma de hacerlo
        # def self.discardElement(usesLeft)
        #     return @@generator.rand(1)<(@@MAX_USES-usesLeft)/(1.0*@@MAX_USES)
        # end
        
        # Genera el siguiente paso que debe de dar en base a los parámetros proporcionados y en base a la probabilidad calculada
        #
        # @param preference [Symbol] El paso que el jugador prefiere realizar.
        # @param validMoves [Array<Symbol>] Los pasos válidos que el jugador puede realizar.
        # @param intelligence [Float] La inteligencia del jugador.
        # @return [Symbol] El siguiente paso que el jugador debe dar.
        def self.nextStep(preference, validMoves, intelligence)
            out = preference

            # if @@generator.rand < intelligence/@@MAX_INTELLIGENCE
            #     out = validMoves[@@generator.rand(validMoves.size)]
            # end

            if randomIntelligence > intelligence
                out = validMoves[@@generator.rand(validMoves.size)]
            end

            out
        end

        

    end
end 